cmd/syncthing/main.go

```diff
@@ -907,6 +901,32 @@ func autoUpgrade(cfg config.Wrapper, app *syncthing.App, evLogger events.Logger)
 	}
 }
 
+func initialAutoUpgrade(misc *db.NamespacedKV) (upgrade.Release, error) {
+	if last, ok, err := misc.Time(upgradeCheckKey); err != nil {
+		return upgrade.Release{}, err
+	} else if ok && time.Since(last) < upgradeCheckInterval {
+		return upgrade.Release{}, errTooEarlyUpgradeCheck
+	}
+	_ = misc.PutTime(upgradeCheckKey, time.Now())
+	release, err := checkUpgrade()
+	if err != nil {
+		return upgrade.Release{}, err
+	}
+	if lastVersion, ok, err := misc.String(upgradeVersionKey); err != nil {
+		return upgrade.Release{}, err
+	} else if ok && lastVersion == release.Tag {
+		// Only check time if we try to upgrade to the same release.
+		if lastTime, ok, err := misc.Time(upgradeTimeKey); err != nil {
+			return upgrade.Release{}, err
+		} else if ok && time.Since(lastTime) < upgradeInterval {
+			return upgrade.Release{}, errTooEarlyUpgrade
+		}
+	}
+	_ = misc.PutString(upgradeVersionKey, release.Tag)
+	_ = misc.PutTime(upgradeTimeKey, time.Now())
+	return release, upgrade.To(release)
```

Comment:
This is a sneakily hidden "oh, and also do the actual upgrade" at the end of the function :) I think this function would be better off as a `checkEarlyUpgrade(misc) (upgradeTo Release, shouldUpgrade bool)` and then the actual upgrade step outside if it returns a release + true?